src/transport.rs

```diff
@@ -0,0 +1,467 @@
+
+        if url.path().is_empty() || url.path() == "/" {
+            return Err("DSN doesn't have a path");
+        }
+
+        match url.host_str() {
+            Some(host) => {
+                let auth = format!(
+                    "Sentry sentry_key={}, sentry_version={}, sentry_client={}",
+                    url.username(),
+                    API_VERSION,
+                    USER_AGENT
+                );
+
+                let uri = format!(
+                    "{}://{}/api/{}/envelope/",
+                    url.scheme(),
+                    host,
+                    &url.path()[1..]
+                );
+
+                Ok(Self { auth, uri })
+            }
+            None => Err("DSN doesn't have a host"),
+        }
+    }
+}
+
+/// Constructs an HTTP request for the provided [`sys::Envelope`] with the DSN
+/// that was registered with the SDK
+fn construct_request(
+    envelope: *mut sys::Envelope,
+    dsn: &Dsn,
+) -> Result<http::Request<Envelope>, ()> {
+    let mut builder = http::Request::builder();
+
+    {
+        let headers = builder.headers_mut().expect("unable to mutate headers");
+        headers.insert("user-agent", USER_AGENT.parse().unwrap());
+        headers.insert("content-type", ENVELOPE_MIME.parse().unwrap());
+        headers.insert("accept", "*/*".parse().unwrap());
+    }
+
+    builder = builder.method("POST");
+    builder = dsn.build_req(builder);
+
+    // Get the DSN for the options, which informs us where to send the request, and what auth token to use
+    let envelope = unsafe {
+        let mut envelope_size = 0;
+        let serialized_envelope = sys::envelope_serialize(envelope, &mut envelope_size);
+
+        if envelope_size == 0 || serialized_envelope.is_null() {
+            return Err(());
+        }
+
+        builder = builder.header("content-length", envelope_size);
+
+        Envelope {
+            inner: envelope,
+            data: serialized_envelope,
+            len: envelope_size,
+        }
+    };
+
+    builder.body(envelope).map_err(|_| ())
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn parses_dsn() {
+        {
+            let hosted =
+                "https://a0b1c2d3e4f5678910abcdeffedcba12@o209016.ingest.sentry.io/0123456";
+
+            let mut builder = http::Request::builder();
+            let dsn: Dsn = hosted.parse().expect("failed to parse hosted DSN");
+            builder = dsn.build_req(builder);
+
+            assert_eq!(
+                builder.uri_ref().unwrap(),
+                "https://o209016.ingest.sentry.io/api/0123456/envelope/"
+            );
+            let headers = builder.headers_ref().unwrap();
+            assert_eq!(headers.get("x-sentry-auth").unwrap(), &format!("Sentry sentry_key=a0b1c2d3e4f5678910abcdeffedcba12, sentry_version={}, sentry_client={}", API_VERSION, USER_AGENT));
+        }
+
+        {
+            let private = "http://a0b1c2d3e4f5678910abcdeffedcba12@192.168.1.1/0123456";
```

Comment:
On a side note, did you get application crashes to work for Sentry on-premise? For me it always errors out on minidump processing and the logs show that the symbolicator fails on an unknown field `name`.

Reply:
We use the hosted service, not sure what the difference would be between it and on-premise, but didn't have an problems that I saw.

Reply:
Alright, thanks.